---
title: 第二十六讲——异常处理
id: 27
date: 2019-8-14 20:00:00
tags: Python
comment: true
---

### 学习大纲

- 异常和错误
- 异常处理
- 补充知识点

<!-----more----->

### 异常和错误

1，错误的种类

1. 语法错误：这种错误在python解释器解释的时候是根本过不了的，必须在程序执行之前进行改正。
2. 逻辑错误：这个是我们今天解决的主要问题。

2，什么是异常

- 代码发生异常错误之后程序中断。

异常的种类 

```python
错误类型 	说明
以下内容必讲 	 
ZeroDivisionError 	除(或取模)零 (所有数据类型)
ValueError 	传入无效的参数
AssertionError 	断言语句失败
StopIteration 	迭代器没有更多的值
IndexError 	序列中没有此索引(index)
IndentationError 	缩进错误
OSError 	输入/输出操作失败
ImportError 	导入模块/对象失败
NameError 	未声明/初始化对象 (没有属性)
AttributeError 	对象没有这个属性
  	 
GeneratorExit 	生成器(generator)发生异常来通知退出
TypeError 	对类型无效的操作
KeyboardInterrupt 	用户中断执行(通常是输入^C)
OverflowError 	数值运算超出最大限制
FloatingPointError 	浮点计算错误
BaseException 	所有异常的基类
SystemExit 	解释器请求退出
Exception 	常规错误的基类
StandardError 	所有的内建标准异常的基类
ArithmeticError 	所有数值计算错误的基类
EOFError 	没有内建输入,到达EOF 标记
EnvironmentError 	操作系统错误的基类
WindowsError 	系统调用失败
LookupError 	无效数据查询的基类
KeyError 	映射中没有这个键
MemoryError 	内存溢出错误(对于Python 解释器不是致命的)
UnboundLocalError 	访问未初始化的本地变量
ReferenceError 	弱引用(Weak reference)试图访问已经垃圾回收了的对象
RuntimeError 	一般的运行时错误
NotImplementedError 	尚未实现的方法
SyntaxError Python 	语法错误
TabError 	Tab 和空格混用
SystemError 	一般的解释器系统错误
UnicodeError 	Unicode 相关的错误
UnicodeDecodeError 	Unicode 解码时的错误
UnicodeEncodeError 	Unicode 编码时错误
UnicodeTranslateError 	Unicode 转换时错误

以下为警告类型 	 
Warning 	警告的基类
DeprecationWarning 	关于被弃用的特征的警告
FutureWarning 	关于构造将来语义会有改变的警告
OverflowWarning 	旧的关于自动提升为长整型(long)的警告
PendingDeprecationWarning 	关于特性将会被废弃的警告
RuntimeWarning 	可疑的运行时行为(runtime behavior)的警告
SyntaxWarning 	可疑的语法的警告
UserWarning 	用户代码生成的警告
```

### 异常处理

1.什么是异常处理

​	1.程序员编写特定的代码，专门用来捕捉这个异常（这段代码与程序逻辑无关，与异常处理有关），如果捕捉成功则进入另外一个处理分支，执行你为其定制的逻辑，使程序不会崩溃，这就是异常处理。

​	2.为什么要进行异常处理

​		1.为了使用户的体验更好。

​		2.增强代码的健壮性和容错性。

2.异常处理的方式

​	1.使用if判断式。
- if判断是的异常处理只能针对某一段代码，对于不同的代码的相同类型的错误需要重新写if来进行处理。

- 在程序中频繁的写与本程序无关，与异常处理有关的if，会使代码的可读性变得非常差。

- if是可以解决异常的，但是万事有好坏之分而已。下面我们去了解另外一种异常处理的方式。

  2.try处理异常

3.**try处理异常的几种类型**

一、单分支

```python
语法：
try:
	被检测的代码块
except 异常类型：
	try中一旦检测到异常，就会执行这个位置的逻辑

try:
	num = int(input("输入数字："))
except ValueError:
	print ("ValueError:输入的非数字!")
#输入数字：ss
#ValueError:输入的非数字!
```

二、多分支

```python
try:
	for i in 123:
		pass
	num = int(input("输入数字："))

except TypeError:
	print ("TypeError:索引非可迭代数据类型！")
except ValueError:
	print ("ValueError:输入的非数字!")

#TypeError:索引非可迭代数据类型！

解释：这段代码中有两个错误，一个是for循环的是一个数字，这是一个typeError的错误；一个是输入的错误，我们执行程序就会在for循环这里报错退出程序了。
```

三、万能异常处理

```python
s1 = 'hello'
try:
    int(s1)
except Exception,e:
    '丢弃或者执行其他逻辑'
    print(e)
#这个万能异常统一使用Exception，没错，是可以捕捉所有的异常，但是这个时候我们就不能将各种错误分开单独的处理了，我们这个时候要么统一将他们丢弃，要么使用同一段代码取处理他们。
```

四、多分支+万能异常处理

```python
#学了上面的多分支和万能异常我们是不是发现了他们各自的不足之处：对于多分支而言我们是可以处理我们知道的错误，但是我们不能控制那些我们未知的错误。而万能错误，我们捕捉到所有的错误，但是不能进行分开的处理。综上，我们就发现了将他们两个合到一起就可以解决我们的需求了。
try:
	num = int(input("输入数字："))
	for i in "aa":
		pass
	s = {12:"asd"}
	print (s["asfsd"])


except ValueError:
	print ("输入的非数字错误。")
except TypeError:
	print ("索引超出范围。")
except Exception as e:
	print (e)
```

五、Try_else_finally

```python
s1 = 'hello'
try:
    int(s1)
except IndexError as e:
    print(e)
except KeyError as e:
    print(e)
except ValueError as e:
    print(e)
#except Exception as e:
#    print(e)
else:
    print('try内代码块没有异常则执行我')
finally:
    print('无论异常与否,都会执行该模块,通常是进行清理工作')
    
#fnally一般用在什么地方：用在关闭数据库，文件句柄的关闭，数据保存等情况。
```

六、Raise:主动触发异常

```python
try:
    raise TypeError('类型错误')
except Exception as e:
    print(e)
```

七、断言

```python
#Assert:表现一种强硬的态度，在assert后面添加一个条件，条件满足就跳过，如果不成立就报错。
a = "pl"
b = input()
assert a == b
#当a和b一旦不相等的时候就会触发断言的错误。
```

八、自定义异常

```python
class EvaException(BaseException):
    def __init__(self,msg):
        self.msg=msg
    def __str__(self):
        return self.msg

try:
    raise EvaException('类型错误')
except EvaException as e:
    print(e)
#为什么会有自定义错误类型？万能异常不是帮我们把所有的错误都捕捉到了吗，其实并不是这样的，python内部有非常多的错误，我们的异常处理是每一种错误类型都是一个类，我们的内部没有把所有恶错误类型都定义进去，所以，有些错误我们需要自己定义。
```

5.try...except方法和if的对比的好处

​	1.把错误处理和真正的工作分开。

​	2.代码更易组织，更清晰，复杂的工作任务更容易实现。

​	3.毫无疑问，更加安全，程序不会因为一点小的错误导致其崩溃。

6.总结：

- try...except应该尽量少用，因为它本身就是你附加给你的程序的一种异常处理的逻辑，与你的主要的工作是没有关系的
  这种东西加的多了，会导致你的代码可读性变差，只有在有些异常无法预知的情况下，才应该加上try...except，其他的逻辑错误应该尽量修正

- 什么时候使用异常处理

  - 什么时候使用异常处理：

    用if代码解决不了的，if代码解决过程繁琐复杂，异常处理慎用，精用。因为try的开销很大的。比如像input就使用try非常好，因为用户的输入是不可控的。

### 	知识点补充

```python
1，所有的错误类型都继承BaseException类
2,语句块中异常发生点后的剩余语句永远不会到达
3，except(异常数据类型一，异常数据类型二.....)，捕获多个异常
4，捕获所有的异常的几种方式
	except
    except Exception
    Exceppt BaseException
5,else:可选的选项，无异常发生执行里面的语句
6，执行流程
	try->异常->except->finally
    try->无异常->else->finally
7，try-finally:用于维持一致性，比如关闭文件
8，sys.exc_info()获取异常信息
```

断言和异常的区别

```
断言用在那些你知道绝对不会发生的事情上，但是因为人总是会犯错误，保不准你写出来的东西跟你想的不一样。所以assert用来捕捉的是程序员自己的错误。（一般使用在程序调式的时候，在生产环境中不会生效）。
assert语句用于检测某个条件表达式是否为真。assert语句又称为断言语句，即assert认为检测的表达式永远为真。

但异常（exception）主要用于捕捉的是用户或者环境的错误

rasie的定义是显示的抛出异常，用户可以使用raise进行判断，显式的引发异常，raise执行后程序将不再向下执行。
```

